build.zig

```diff
@@ -22,23 +21,114 @@ pub fn build(b: *std.Build) !void {
+    install: *std.build.Step.InstallArtifact,
+
+    pub fn init(b: *std.Build, options: struct {
+        name: []const u8,
+        src: []const u8,
+        optimize: std.builtin.OptimizeMode,
+        deps: []const Dependency = &.{},
+    }) !App {
+        b.lib_dir = b.fmt("{s}/www", .{b.install_path});
+
+        const wasm32_freestanding = std.zig.CrossTarget{ .cpu_arch = .wasm32, .os_tag = .freestanding };
+        const lib = b.addSharedLibrary(.{
+            .name = options.name,
+            .root_source_file = .{ .path = options.src },
+            .optimize = options.optimize,
+            .target = wasm32_freestanding,
+        });
+        lib.rdynamic = true;
+
+        b.installArtifact(lib);
+        const install = b.addInstallArtifact(lib, .{});
+
+        var inits = std.ArrayList(u8).init(b.allocator);
+        var imports = std.ArrayList(u8).init(b.allocator);
+        const import_writer = imports.writer();
+        for (options.deps) |dep| {
+            try import_writer.print(
+                "import * as {s} from './{s}';\n",
+                .{ dep.name, std.fs.path.basename(dep.binding_file) },
+            );
+
+            try inits.writer().print("{s}.init(wasm);\n", .{dep.name});
+
+            const module = b.createModule(.{
+                .source_file = .{ .path = b.pathFromRoot(dep.wrapper_file) },
+            });
+            lib.addModule(dep.name, module);
+
+            const install_js = b.addInstallFile(
+                .{ .path = dep.binding_file },
+                try std.fs.path.join(b.allocator, &.{ "www/", std.fs.path.basename(dep.binding_file) }),
+            );
+            install.step.dependOn(&install_js.step);
+        }
+
+        try import_writer.writeAll("let imports = {\nenv: {\n");
+        for (options.deps) |dep| {
+            try import_writer.print("...{s},\n", .{dep.name});
+        }
+        try import_writer.writeAll("}\n};\n");
+
+        // TODO: make the loader configurable
+        const install_loader = b.addInstallFile(
+            .{ .path = "www/loader.js" },
+            try std.fs.path.join(b.allocator, &.{ "www", "loader.js" }),
+        );
+        install.step.dependOn(&install_loader.step);
+
+        // TODO: do the format at runtime
+        const index_formatted = try std.fmt.allocPrint(b.allocator, index_html, .{
+            .app_name = options.name,
+            .wasm_path = options.name,
+            .imports = imports.items,
+            .inits = inits.items,
+        });
+
+        const write_index_html = b.addWriteFile(b.getInstallPath(.lib, "index.html"), index_formatted);
+        install.step.dependOn(&write_index_html.step);
+
+        return .{
+            .b = b,
+            .name = options.name,
+            .compile = lib,
+            .install = install,
+        };
+    }
+
+    pub const Dependency = struct {
+        name: []const u8,
+        binding_file: []const u8,
+        wrapper_file: []const u8,
```

Comment:
"binding" and "wrapper" are almost equal words for me, so it's impossible for me to tell from the names which of these is supposed to be which.

Consider using the same naming as in `generateFiles` instead:

```
    generated_sysjs_zig: []const u8,
    generated_sysjs_js: []const u8,
```

Reply:
Completely agree